src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/action/SearchIssues.kt

```diff
@@ -0,0 +1,133 @@
+    private val jqlMemory: JqlMemory,
+    private val issueKeyMemory: IssueKeyMemory,
+    private val desiredView: IssueNavResultsView
+) : Action {
+
+    private val logger: Logger = LogManager.getLogger(this::class.java)
+
+    override fun run() {
+        val jqlQuery = jqlMemory.recall()
+        if (jqlQuery == null) {
+            logger.debug("Skipping ${actionType.label} action, because I cannot recall any JQL queries.")
+            return
+        }
+        meter.measure(
+            key = actionType,
+            action = { search(jqlQuery) },
+            observation = this::observe
+        )
+    }
+
+    private fun search(jqlQuery: String) = jira
+        .goToIssueNavigator(jqlQuery)
+        .also { switchResultsView() }
+        .waitForResults(desiredView)
+
+    private fun switchResultsView() {
+        meter.measure(
+            switchActionType,
+            {
+                val observation = SwitchObservation.Builder(desiredView.javaClass.simpleName)
+                val selectedBefore = desiredView.isSelected()
+                observation.selectedBefore(selectedBefore)
+                if (!selectedBefore) {
+                    desiredView.switchToView()
+                    observation.switched()
+                }
+                observation
+            },
+            { observation ->
+                observation
+                    .selectedAfter(desiredView.isSelected())
+                    .build()
+                    .serialize()
+            }
+        )
+    }
+
+    private fun observe(page: IssueNavigatorPage): JsonObject {
+        val issueKeys = desiredView.listIssueKeys()
+        issueKeyMemory.remember(issueKeys)
+        return SearchJqlObservation(page.jql, issueKeys.size, desiredView.countResults() ?: -1).serialize()
+    }
+
+    class Builder(
+        private var jira: WebJira,
+        private var meter: ActionMeter,
+        seededRandom: SeededRandom
+    ) {
+
+        private var actionType: ActionType<SearchJqlObservation> = SEARCH_WITH_JQL
+        private var jqlMemory: JqlMemory = AdaptiveJqlMemory(seededRandom)
+        private var issueKeyMemory: IssueKeyMemory = AdaptiveIssueKeyMemory(seededRandom)
+        private var desiredView: IssueNavResultsView = DetailView(jira.driver)
+
+        fun actionType(actionType: ActionType<SearchJqlObservation>) = apply { this.actionType = actionType }
+        fun jqlMemory(jqlMemory: JqlMemory) = apply { this.jqlMemory = jqlMemory }
+        fun issueKeyMemory(issueKeyMemory: IssueKeyMemory) = apply { this.issueKeyMemory = issueKeyMemory }
+        fun desiredView(desiredView: IssueNavResultsView) = apply { this.desiredView = desiredView }
+
+        fun build(): Action = SearchIssues(jira, meter, actionType, jqlMemory, issueKeyMemory, desiredView)
+    }
+}
+
+private val switchActionType = ActionType("Switch issue nav view") { SwitchObservation(it) }
+
+private class SwitchObservation(
```

Comment:
If this is a private entity under scope of `SearchIssues` (it's in the `SearchIssues.kt` file), shouldn't it be inside the `SearchIssues` class?